File: python/x402/tests/mocks/cash.py
"""Mock "cash" scheme implementation for testing.

This provides a simple mock payment scheme that simulates cash payments
for testing the core x402Client, x402ResourceServer, and x402Facilitator
integration without needing real blockchain infrastructure.
"""

import time
from typing import Any

from x402.facilitator import x402Facilitator, x402FacilitatorSync
from x402.schemas import (
    AssetAmount,
    Network,
    PaymentPayload,
    PaymentRequirements,
    Price,
    SettleResponse,
    SupportedKind,
    SupportedResponse,
    VerifyResponse,
)


class CashSchemeNetworkClient:
    """Client-side mock "cash" scheme implementation.

    Creates payment payloads with a simple signature format.

    Attributes:
        scheme: The scheme identifier ("cash").
    """

    scheme = "cash"

    def __init__(self, payer: str) -> None:
        """Create a CashSchemeNetworkClient.

        Args:
            payer: The name of the payer (used in signature).
        """
        self._payer = payer

    def create_payment_payload(
        self,
        requirements: PaymentRequirements,
    ) -> dict[str, Any]:
        """Create a cash payment payload.

        The payload contains:
        - signature: ~{payer_name}
        - validUntil: timestamp when payment expires
        - name: payer's name

        Args:
            requirements: The payment requirements.

        Returns:
            Inner payload dict with signature and validity.
        """
        valid_until = int(time.time() * 1000) + (requirements.max_timeout_seconds * 1000)
        return {
            "signature": f"~{self._payer}",
            "validUntil": str(valid_until),
            "name": self._payer,
        }


class CashSchemeNetworkFacilitator:
    """Facilitator-side mock "cash" scheme implementation.

    Verifies cash payments by checking signature format and expiration.

    Attributes:
        scheme: The scheme identifier ("cash").
        caip_family: The CAIP family pattern ("x402:*").
    """

    scheme = "cash"
    caip_family = "x402:*"

    def get_extra(self, network: Network) -> dict[str, Any] | None:
        """Get extra data for supported kinds.

        Args:
            network: The network identifier.

        Returns:
            Empty dict for cash scheme.
        """
        return {}

    def get_signers(self, network: Network) -> list[str]:
        """Get signer addresses.

        Args:
            network: The network identifier.

        Returns:
            Empty list for cash scheme.
        """
        return []

    def verify(
        self,
        payload: PaymentPayload,
        requirements: PaymentRequirements,
        context=None,
    ) -> VerifyResponse:
        """Verify a cash payment.

        Checks:
        1. Signature matches expected format (~{name})
        2. Payment hasn't expired

        Args:
            payload: The payment payload.
            requirements: The payment requirements.
            context: Optional facilitator context.

        Returns:
            VerifyResponse with is_valid=True or False.
        """
        inner = payload.payload

        # Check signature format
        expected_sig = f"~{inner.get('name', '')}"
        if inner.get("signature") != expected_sig:
            return VerifyResponse(
                is_valid=False,
                invalid_reason="invalid_signature",
                payer=None,
            )

        # Check expiration
        valid_until = int(inner.get("validUntil", "0"))
        if valid_until < int(time.time() * 1000):
            return VerifyResponse(
                is_valid=False,
                invalid_reason="expired_signature",
                payer=None,
            )

        return VerifyResponse(
            is_valid=True,
            invalid_reason=None,
            payer=inner.get("signature"),
        )

    def settle(
        self,
        payload: PaymentPayload,
        requirements: PaymentRequirements,
        context=None,
    ) -> SettleResponse:
        """Settle a cash payment.

        First verifies the payment, then creates a settlement transaction.

        Args:
            payload: The payment payload.
            requirements: The payment requirements.

        Returns:
            SettleResponse with success=True or False.
        """
        verify_result = self.verify(payload, requirements, context)

        if not verify_result.is_valid:
            return SettleResponse(
                success=False,
                error_reason=verify_result.invalid_reason,
                payer=verify_result.payer,
                transaction="",
                network=requirements.network,
            )

        name = payload.payload.get("name", "Unknown")
        transaction = (
            f"{name} transferred {requirements.amount} "
            f"{requirements.asset} to {requirements.pay_to}"
        )

        return SettleResponse(
            success=True,
            error_reason=None,
            payer=payload.payload.get("signature"),
            transaction=transaction,
            network=requirements.network,
        )


class CashSchemeNetworkServer:
    """Server-side mock "cash" scheme implementation.

    Parses prices and enhances payment requirements.

    Attributes:
        scheme: The scheme identifier ("cash").
    """

    scheme = "cash"

    def parse_price(self, price: Price, network: Network) -> AssetAmount:
        """Parse a price into AssetAmount.

        Handles:
        - AssetAmount objects (passthrough)
        - String prices like "$10" or "10 USD"
        - Numeric prices

        Args:
            price: The price to parse.
            network: The network identifier.

        Returns:
            AssetAmount with amount and asset.
        """
        # Handle AssetAmount object
        if isinstance(price, AssetAmount):
            return AssetAmount(
                amount=price.amount,
                asset=price.asset or "USD",
                extra={},
            )

        # Handle string prices
        if isinstance(price, str):
            clean_price = price.lstrip("$").rstrip(" USD").rstrip(" usd").strip()
            return AssetAmount(
                amount=clean_price,
                asset="USD",
                extra={},
            )

        # Handle numeric prices
        if isinstance(price, int | float):
            return AssetAmount(
                amount=str(price),
                asset="USD",
                extra={},
            )

        raise ValueError(f"Invalid price format: {price}")

    def enhance_payment_requirements(
        self,
        requirements: PaymentRequirements,
        supported_kind: SupportedKind,
        extensions: list[str],
    ) -> PaymentRequirements:
        """Enhance payment requirements.

        Cash scheme doesn't need any enhancements.

        Args:
            requirements: Base payment requirements.
            supported_kind: The supported kind from facilitator.
            extensions: Extension keys.

        Returns:
            Unmodified payment requirements.
        """
        return requirements


class CashFacilitatorClient:
    """Mock FacilitatorClient that wraps an x402Facilitator (async).

    Used for testing the server integration without HTTP.

    Attributes:
        scheme: The scheme identifier ("cash").
        network: The network identifier ("x402:cash").
    """

    scheme = "cash"
    network = "x402:cash"
    x402_version = 2

    def __init__(self, facilitator: x402Facilitator) -> None:
        """Create a CashFacilitatorClient.

        Args:
            facilitator: The x402Facilitator to wrap.
        """
        self._facilitator = facilitator

    async def verify(
        self,
        payload: PaymentPayload,
        requirements: PaymentRequirements,
    ) -> VerifyResponse:
        """Verify a payment through the facilitator.

        Args:
            payload: The payment payload.
            requirements: The payment requirements.

        Returns:
            VerifyResponse from the facilitator.
        """
        return await self._facilitator.verify(payload, requirements)

    async def settle(
        self,
        payload: PaymentPayload,
        requirements: PaymentRequirements,
    ) -> SettleResponse:
        """Settle a payment through the facilitator.

        Args:
            payload: The payment payload.
            requirements: The payment requirements.

        Returns:
            SettleResponse from the facilitator.
        """
        return await self._facilitator.settle(payload, requirements)

    def get_supported(self) -> SupportedResponse:
        """Get supported payment kinds.

        Returns:
            SupportedResponse with cash scheme support.
        """
        return SupportedResponse(
            kinds=[
                SupportedKind(
                    x402_version=self.x402_version,
                    scheme=self.scheme,
                    network=self.network,
                    extra={},
                )
            ],
            extensions=[],
            signers={},
        )


class CashFacilitatorClientSync:
    """Mock FacilitatorClient that wraps an x402FacilitatorSync.

    Used for testing the server integration without HTTP.

    Attributes:
        scheme: The scheme identifier ("cash").
        network: The network identifier ("x402:cash").
    """

    scheme = "cash"
    network = "x402:cash"
    x402_version = 2

    def __init__(self, facilitator: x402FacilitatorSync) -> None:
        """Create a CashFacilitatorClientSync.

        Args:
            facilitator: The x402FacilitatorSync to wrap.
        """
        self._facilitator = facilitator

    def verify(
        self,
        payload: PaymentPayload,
        requirements: PaymentRequirements,
    ) -> VerifyResponse:
        """Verify a payment through the facilitator.

        Args:
            payload: The payment payload.
            requirements: The payment requirements.

        Returns:
            VerifyResponse from the facilitator.
        """
        return self._facilitator.verify(payload, requirements)

    def settle(
        self,
        payload: PaymentPayload,
        requirements: PaymentRequirements,
    ) -> SettleResponse:
        """Settle a payment through the facilitator.

        Args:
            payload: The payment payload.
            requirements: The payment requirements.

        Returns:
            SettleResponse from the facilitator.
        """
        return self._facilitator.settle(payload, requirements)

    def get_supported(self) -> SupportedResponse:
        """Get supported payment kinds.

        Returns:
            SupportedResponse with cash scheme support.
        """
        return SupportedResponse(
            kinds=[
                SupportedKind(
                    x402_version=self.x402_version,
                    scheme=self.scheme,
                    network=self.network,
                    extra={},
                )
            ],
            extensions=[],
            signers={},
        )


def build_cash_payment_requirements(
    pay_to: str,
    asset: str,
    amount: str,
) -> PaymentRequirements:
    """Build payment requirements for the cash scheme.

    Args:
        pay_to: The recipient address/name.
        asset: The asset being paid (e.g., "USD").
        amount: The amount to pay.

    Returns:
        PaymentRequirements for cash scheme.
    """
    return PaymentRequirements(
        scheme="cash",
        network="x402:cash",
        asset=asset,
        amount=amount,
        pay_to=pay_to,
        max_timeout_seconds=1000,
        extra={},
    )

File: go/test/mocks/cash/cash.go
package cash

import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"

	x402 "github.com/x402-foundation/x402/go"
	"github.com/x402-foundation/x402/go/types"
)

// ============================================================================
// Cash Scheme Network Client
// ============================================================================

// SchemeNetworkClient implements the client side of the cash payment scheme
type SchemeNetworkClient struct {
	payer string
}

// NewSchemeNetworkClient creates a new cash scheme client
func NewSchemeNetworkClient(payer string) *SchemeNetworkClient {
	return &SchemeNetworkClient{
		payer: payer,
	}
}

// Scheme returns the payment scheme identifier
func (c *SchemeNetworkClient) Scheme() string {
	return "cash"
}

// CreatePaymentPayload creates a V2 payment payload for the cash scheme
func (c *SchemeNetworkClient) CreatePaymentPayload(ctx context.Context, requirements types.PaymentRequirements) (types.PaymentPayload, error) {
	validUntil := time.Now().Add(time.Duration(requirements.MaxTimeoutSeconds) * time.Second).Unix()

	return types.PaymentPayload{
		X402Version: 2,
		Payload: map[string]interface{}{
			"signature":  fmt.Sprintf("~%s", c.payer),
			"validUntil": strconv.FormatInt(validUntil, 10),
			"name":       c.payer,
		},
	}, nil
}

// ============================================================================
// Cash Scheme Network Facilitator
// ============================================================================

// SchemeNetworkFacilitator implements the facilitator side of the cash payment scheme
type SchemeNetworkFacilitator struct{}

// NewSchemeNetworkFacilitator creates a new cash scheme facilitator
func NewSchemeNetworkFacilitator() *SchemeNetworkFacilitator {
	return &SchemeNetworkFacilitator{}
}

// Scheme returns the payment scheme identifier
func (f *SchemeNetworkFacilitator) Scheme() string {
	return "cash"
}

// CaipFamily returns the CAIP family pattern
func (f *SchemeNetworkFacilitator) CaipFamily() string {
	return "x402:*"
}

// GetExtra returns mechanism-specific extra data for the supported kinds endpoint.
// For the mock cash scheme, return nil.
func (f *SchemeNetworkFacilitator) GetExtra(_ x402.Network) map[string]interface{} {
	return nil
}

// GetSigners returns signer addresses
func (f *SchemeNetworkFacilitator) GetSigners(_ x402.Network) []string {
	return []string{}
}

// Verify verifies a V2 payment payload against requirements (typed)
func (f *SchemeNetworkFacilitator) Verify(ctx context.Context, payload types.PaymentPayload, requirements types.PaymentRequirements, _ *x402.FacilitatorContext) (*x402.VerifyResponse, error) {
	// Extract payload fields
	signature, ok := payload.Payload["signature"].(string)
	if !ok {
		return nil, x402.NewVerifyError("missing_signature", "", "missing signature")
	}

	name, ok := payload.Payload["name"].(string)
	if !ok {
		return nil, x402.NewVerifyError("missing_name", "", "missing name")
	}

	validUntilStr, ok := payload.Payload["validUntil"].(string)
	if !ok {
		return nil, x402.NewVerifyError("missing_validUntil", "", "missing validUntil")
	}

	// Check signature
	expectedSig := fmt.Sprintf("~%s", name)
	if signature != expectedSig {
		return nil, x402.NewVerifyError("invalid_signature", signature, fmt.Sprintf("invalid signature: %s != %s", signature, expectedSig))
	}

	// Check expiration
	validUntil, err := strconv.ParseInt(validUntilStr, 10, 64)
	if err != nil {
		return nil, x402.NewVerifyError("invalid_validUntil", signature, fmt.Sprintf("invalid validUntil: %s", validUntilStr))
	}

	if validUntil < time.Now().Unix() {
		return nil, x402.NewVerifyError("expired_signature", signature, fmt.Sprintf("expired signature: %d < %d", validUntil, time.Now().Unix()))
	}

	return &x402.VerifyResponse{
		IsValid: true,
		Payer:   signature,
	}, nil
}

// Settle settles a V2 payment (typed)
func (f *SchemeNetworkFacilitator) Settle(ctx context.Context, payload types.PaymentPayload, requirements types.PaymentRequirements, fctx *x402.FacilitatorContext) (*x402.SettleResponse, error) {
	network := x402.Network(requirements.Network)

	// First verify the payment
	verifyResponse, err := f.Verify(ctx, payload, requirements, fctx)
	if err != nil {
		// Convert VerifyError to SettleError
		var ve *x402.VerifyError
		if errors.As(err, &ve) {
			return nil, x402.NewSettleError(ve.InvalidReason, ve.Payer, network, "", ve.InvalidMessage)
		}
		return nil, x402.NewSettleError("verification_failed", "", network, "", err.Error())
	}

	// Extract name for transaction message
	name, _ := payload.Payload["name"].(string)

	return &x402.SettleResponse{
		Success:     true,
		Transaction: fmt.Sprintf("%s transferred %s %s to %s", name, requirements.Amount, requirements.Asset, requirements.PayTo),
		Network:     network,
		Payer:       verifyResponse.Payer,
	}, nil
}

// ============================================================================
// Cash Scheme Network Server
// ============================================================================

// SchemeNetworkServer implements the server side of the cash payment scheme
type SchemeNetworkServer struct{}

// NewSchemeNetworkServer creates a new cash scheme server
func NewSchemeNetworkServer() *SchemeNetworkServer {
	return &SchemeNetworkServer{}
}

// Scheme returns the payment scheme identifier
func (s *SchemeNetworkServer) Scheme() string {
	return "cash"
}

// ParsePrice parses a price into asset amount format
func (s *SchemeNetworkServer) ParsePrice(price x402.Price, network x402.Network) (x402.AssetAmount, error) {
	// Handle pre-parsed price object
	if assetAmount, ok := price.(x402.AssetAmount); ok {
		return assetAmount, nil
	}

	// Handle map format
	if priceMap, ok := price.(map[string]interface{}); ok {
		amount, _ := priceMap["amount"].(string)
		asset, _ := priceMap["asset"].(string)
		if asset == "" {
			asset = "USD"
		}
		return x402.AssetAmount{
			Amount: amount,
			Asset:  asset,
			Extra:  nil,
		}, nil
	}

	// Parse string prices like "$10" or "10 USD"
	if priceStr, ok := price.(string); ok {
		// Remove dollar sign and USD suffix
		cleanPrice := strings.TrimPrefix(priceStr, "$")
		cleanPrice = strings.TrimSuffix(cleanPrice, " USD")
		cleanPrice = strings.TrimSuffix(cleanPrice, "USD")
		cleanPrice = strings.TrimSpace(cleanPrice)

		return x402.AssetAmount{
			Amount: cleanPrice,
			Asset:  "USD",
			Extra:  nil,
		}, nil
	}

	// Handle number input
	if priceNum, ok := price.(float64); ok {
		return x402.AssetAmount{
			Amount: fmt.Sprintf("%.2f", priceNum),
			Asset:  "USD",
			Extra:  nil,
		}, nil
	}

	if priceInt, ok := price.(int); ok {
		return x402.AssetAmount{
			Amount: strconv.Itoa(priceInt),
			Asset:  "USD",
			Extra:  nil,
		}, nil
	}

	return x402.AssetAmount{}, fmt.Errorf("invalid price format: %v", price)
}

// EnhancePaymentRequirements enhances payment requirements with cash-specific details
func (s *SchemeNetworkServer) EnhancePaymentRequirements(
	ctx context.Context,
	requirements types.PaymentRequirements,
	supportedKind types.SupportedKind,
	facilitatorExtensions []string,
) (types.PaymentRequirements, error) {
	// Cash scheme doesn't need any special enhancements
	return requirements, nil
}

// ============================================================================
// Cash Facilitator Client
// ============================================================================

// FacilitatorClient wraps a facilitator for the cash scheme
type FacilitatorClient struct {
	facilitator *x402.X402Facilitator
}

// NewFacilitatorClient creates a new cash facilitator client
func NewFacilitatorClient(facilitator *x402.X402Facilitator) *FacilitatorClient {
	return &FacilitatorClient{
		facilitator: facilitator,
	}
}

// Verify verifies a payment payload against requirements
func (c *FacilitatorClient) Verify(ctx context.Context, payloadBytes []byte, requirementsBytes []byte) (*x402.VerifyResponse, error) {
	// Pass bytes directly to facilitator (it will unmarshal internally)
	return c.facilitator.Verify(ctx, payloadBytes, requirementsBytes)
}

// Settle settles a payment based on the payload and requirements
func (c *FacilitatorClient) Settle(ctx context.Context, payloadBytes []byte, requirementsBytes []byte) (*x402.SettleResponse, error) {
	// Pass bytes directly to facilitator (it will unmarshal internally)
	return c.facilitator.Settle(ctx, payloadBytes, requirementsBytes)
}

// GetSupported gets supported payment kinds and extensions
func (c *FacilitatorClient) GetSupported(ctx context.Context) (x402.SupportedResponse, error) {
	return x402.SupportedResponse{
		Kinds: []x402.SupportedKind{
			{
				X402Version: 2,
				Scheme:      "cash",
				Network:     "x402:cash",
				Extra:       nil,
			},
		},
		Extensions: []string{},
		Signers:    make(map[string][]string),
	}, nil
}

// Identifier returns the identifier for this facilitator client
func (c *FacilitatorClient) Identifier() string {
	return "cash-facilitator"
}

// ============================================================================
// Helper Functions
// ============================================================================

// BuildPaymentRequirements creates a payment requirements object for the cash scheme
func BuildPaymentRequirements(payTo string, asset string, amount string) types.PaymentRequirements {
	return types.PaymentRequirements{
		Scheme:            "cash",
		Network:           "x402:cash",
		Asset:             asset,
		Amount:            amount,
		PayTo:             payTo,
		MaxTimeoutSeconds: 1000,
		Extra:             nil,
	}
}
